src/core/reactor.cc

```diff
@@ -4746,7 +4746,33 @@ std::chrono::nanoseconds reactor::total_steal_time() {
     // unexpectedly blocked (since CPU time won't tick up when that occurs).
     //
     // But what we have here should be good enough and at least has a well defined meaning.
-    return total_awake_time() - total_cpu_time();
+    //
+    // Because we calculate sleep time with timestamps around polling methods that may sleep, like
+    // io_getevents, we systematically over-count sleep time, since there is CPU usage within the
+    // period timed as sleep, before and after an actual sleep occurs (and no sleep may occur at all,
+    // e.g., if there are events immediately available). Over-counting sleep means we under-count the 
+    // wall-clock awake time, and so if there is no "true" steal, we will generally have a small
+    // *negative* steal time, because we under-count awake wall clock time while thread CPU time does
+    // not have a corresponding error.
+    //
+    // Becuase we claim "steal" is a counter, we must ensure that it never deceases, because PromQL
+    // functions which use counters will produce non-sensical results if they do. Therefore we clamp
+    // the output such that it never decreases.
+    //
+    // Finally, we don't just clamp difference of awake and CPU time since proces start at 0, but
+    // take the last value we returned from this function and then calculate the incremental steal
+    // time since that measurement, clamped to 0. This means that as soon as steal time becomes 
+    // positive, it will be reflected in the measurement, rather than needing to "consume" all the
+    // accumulated negative steal time before positive steal times start showing up.
+
+
+    auto true_steal = total_awake_time() - total_cpu_time();
+    auto mono_steal = _last_mono_steal + std::max(true_steal - _last_true_steal, 0ns);
+
+    _last_true_steal = true_steal;
+    _last_mono_steal = mono_steal;
```

Comment:
In other words, the mono_steal is the sum of all positive changes in true_steal, and the "change" is calculated only at times when this function is called, right?

Reply:
Yes, exactly. It does introduce an "observation effect" in some sense: the results may be different depending on how often you observe the metric, but it's the best I could come up with and seems like it works well in practice since at the observation points are where you care about this.

Reply:
That's unfortunate :( Let's try to keep this "update rate" under some control. I can imagine two ways, none is perfect either.

First -- together with this place, update steal time (or maybe mono-steal part) when reactor stops sleeping.
Second -- arm a (lowres?) timer that will update the steal time, this getter only reports what had been accumulated so far.